text/0000-macro-naming.md

```diff
@@ -0,0 +1,165 @@
+
+The grammar for macros is changed, anywhere we currently parser `name "!"`, we
+now parse `path "!"`. I don't think this introduces any issues.
+
+Name lookup follows the same name resolution rules as other items. See [RFC
+1560](https://github.com/rust-lang/rfcs/pull/1560) for details on how name
+resolution could be adapted to support this.
+
+## Attribute-like macro uses
+
+Attribute macros may also be named using a `::`-separated path. Other than
+appearing in an attribute, these also follow the usual Rust naming rules.
+
+E.g., `#[::foo::bar::baz(...)]` and `#[bar::baz(...)]` are uses of absolute and
+relative paths, respectively.
+
+
+## Importing macros
+
+Importing macros is done using `use` in the same way as other items. An `!` is
+not necessary in an import item. Macros are imported into their own namespace
+and do not shadow or overlap items with the same name in the type or value
+namespaces.
+
+E.g., `use foo::bar::baz;` imports the macro `baz` from the module `::foo::bar`.
+Macro imports may be used in import lists (with other macro imports and with
+non-macro imports).
+
+Where a glob import (`use ...::*;`) imports names from a module including macro
+definitions, the names of those macros are also imported. E.g., `use
+foo::bar::*;` would import `baz` along with any other items in `foo::bar`.
+
+Where macros are defined in a separate crate, these are imported in the same way
+as other items by an `extern crate` item.
+
+No `#[macro_use]` or `#[macro_export]` annotations are required.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+If the new macro system is not well adopted by users, we could be left with two
+very different schemes for naming macros depending on whether a macro is defined
+by example or procedurally. That would be inconsistent and annoying. However, I
+hope we can make the new macro system appealing enough and close enough to the
+existing system that migration is both desirable and easy.
+
+
+# Alternatives
+[alternatives]: #alternatives
+
+We could adopt the proposed scheme for procedural macros only and keep the
+existing scheme for macros by example.
+
+We could adapt the current macros by example scheme to procedural macros.
+
+We could require the `!` in macro imports to distinguish them from other names.
+I don't think this is necessary or helpful.
+
+We could continue to require `macro_export` annotations on top of this scheme.
+However, I prefer moving to a scheme using the same privacy system as the rest
+of Rust, see below.
+
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+## Privacy for macros
+
+I would like that macros follow the same rules for privacy as other Rust items,
+i.e., they are private by default and may be marked as `pub` to make them
+public. This is not as straightforward as it sounds as it requires parsing `pub
+macro_rules! foo` as a macro definition, etc. I leave this for a separate RFC.
```

Comment:
This would need to be solved before stabilization, as it's not backwards compatible to make every `macro_rules` definition private by default.